telegram/ext/builders.py

```diff
@@ -0,0 +1,1338 @@
+            return self._updater_class(
+                dispatcher=dispatcher,
+                user_signal_handler=self._user_signal_handler,
+                exception_event=dispatcher.exception_event,
+                builder_flag=True,
+                **self._updater_kwargs,
+            )
+        return self._updater_class(
+            dispatcher=self._dispatcher,
+            bot=self._dispatcher.bot if self._dispatcher else (self._bot or self._build_ext_bot()),
+            update_queue=self._update_queue,
+            user_signal_handler=self._user_signal_handler,
+            exception_event=self._dispatcher.exception_event
+            if self._dispatcher
+            else self._exception_event,
+            builder_flag=True,
+            **self._updater_kwargs,
+        )
+
+    @staticmethod
+    def _exception_builder(arg_1: str, arg_2: str = None) -> RuntimeError:
+        if not arg_2:
+            return RuntimeError(f'The parameter `{arg_1}` was already set.')
+        return RuntimeError(f'The parameter `{arg_1}` can only be set, if the no {arg_2} was set.')
+
+    @property
+    def _dispatcher_check(self) -> bool:
+        return self._dispatcher_was_set and self._dispatcher is not None
+
+    @check_if_already_set
+    def _set_dispatcher_class(
+        self: BuilderType, dispatcher_class: Type[Dispatcher], kwargs: Dict[str, object] = None
+    ) -> BuilderType:
+        if self._dispatcher_was_set:
+            raise self._exception_builder('dispatcher_class', 'Dispatcher instance')
+        self._dispatcher_class = dispatcher_class
+        self._dispatcher_kwargs = kwargs or {}
+        self._dispatcher_class_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_updater_class(
+        self: BuilderType, updater_class: Type[Updater], kwargs: Dict[str, object] = None
+    ) -> BuilderType:
+        self._updater_class = updater_class
+        self._updater_kwargs = kwargs or {}
+        self._updater_class_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_token(self: BuilderType, token: str) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('token', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('token', 'Dispatcher instance')
+        self._token = token
+        self._token_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_base_url(self: BuilderType, base_url: str) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('base_url', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('base_url', 'Dispatcher instance')
+        self._base_url = base_url
+        self._base_url_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_base_file_url(self: BuilderType, base_file_url: str) -> BuilderType:
+        if self._bot_was_set:
```

Comment:
Should we assert that `self._request` wasn't set as well?

Reply:
https://github.com/python-telegram-bot/python-telegram-bot/pull/2646#discussion_r706642955